File: hindsight-docs/examples/api/memory-banks.py
#!/usr/bin/env python3
"""
Memory Banks API examples for Hindsight.
Run: python examples/api/memory-banks.py
"""
import os
import requests

HINDSIGHT_URL = os.getenv("HINDSIGHT_API_URL", "http://localhost:8888")

# =============================================================================
# Setup (not shown in docs)
# =============================================================================
from hindsight_client import Hindsight

client = Hindsight(base_url=HINDSIGHT_URL)

# =============================================================================
# Doc Examples
# =============================================================================

# [docs:create-bank]
client.create_bank(bank_id="my-bank")
# [/docs:create-bank]


# [docs:bank-with-disposition]
client.create_bank(bank_id="architect-bank")
client.update_bank_config(
    "architect-bank",
    reflect_mission="You're a senior software architect - keep track of system designs, "
            "technology decisions, and architectural patterns. Prefer simplicity over cutting-edge.",
    disposition_skepticism=4,   # Questions new technologies
    disposition_literalism=4,   # Focuses on concrete specs
    disposition_empathy=2,      # Prioritizes technical facts
)
# [/docs:bank-with-disposition]


# [docs:bank-background]
client.create_bank(bank_id="my-bank")
client.update_bank_config(
    "my-bank",
    reflect_mission="I am a research assistant specializing in machine learning.",
)
# [/docs:bank-background]


# [docs:bank-mission]
client.create_bank(bank_id="my-bank")
client.update_bank_config(
    "my-bank",
    reflect_mission="You're a senior software architect - keep track of system designs, "
            "technology decisions, and architectural patterns.",
)
# [/docs:bank-mission]


# [docs:bank-support-agent]
client.create_bank(bank_id="support-bank")
client.update_bank_config(
    "support-bank",
    observations_mission="I am a customer support agent. Track customer preferences, "
            "recurring issues, and resolution history to provide consistent, personalized support.",
)
# [/docs:bank-support-agent]


# [docs:update-bank-config]
client.update_bank_config(
    "my-bank",
    retain_mission="Always include technical decisions, API design choices, and architectural trade-offs. Ignore meeting logistics and social exchanges.",
    retain_extraction_mode="verbose",
    observations_mission="Observations are stable facts about people and projects. Always include preferences, skills, and recurring patterns. Ignore one-off events.",
    disposition_skepticism=4,
    disposition_literalism=4,
    disposition_empathy=2,
)
# [/docs:update-bank-config]


# [docs:get-bank-config]
# Returns resolved config (server defaults merged with bank overrides) and the raw overrides
data = client.get_bank_config("my-bank")
# data["config"]     — full resolved configuration
# data["overrides"]  — only fields overridden at the bank level
# [/docs:get-bank-config]


# [docs:reset-bank-config]
# Remove all bank-level overrides, reverting to server defaults
client.reset_bank_config("my-bank")
# [/docs:reset-bank-config]


# =============================================================================
# Prompt preview and bank transfer (async: the low-level APIs are async-only)
# =============================================================================
import asyncio

TRANSFER_BANKS = ["transfer-py", "transfer-py-copy", "transfer-py-other", "transfer-py-clone"]


async def wait_for(client, bank_id: str, operation_id: str) -> None:
    for _ in range(120):
        status = await client.operations.get_operation_status(bank_id, operation_id)
        if status.status == "completed":
            return
        if status.status in ("failed", "cancelled"):
            raise RuntimeError(f"operation {operation_id} {status.status}: {status.error_message}")
        await asyncio.sleep(1)
    raise TimeoutError(f"operation {operation_id} did not finish")


async def transfer_examples() -> None:
    # A fresh client: the sync calls above ran on another event loop.
    client = Hindsight(base_url=HINDSIGHT_URL)
    for bank_id in TRANSFER_BANKS:
        requests.delete(f"{HINDSIGHT_URL}/v1/default/banks/{bank_id}")
    # Chunks mode stores text verbatim, so no LLM call is needed.
    await client.acreate_bank(bank_id="transfer-py")
    await client.aupdate_bank_config("transfer-py", retain_extraction_mode="chunks")
    await client.aretain_batch(
        "transfer-py",
        [
            {"content": "Alice leads the payments team.", "document_id": "doc-1"},
            {"content": "Bob moved to the Berlin office.", "document_id": "doc-2"},
        ],
    )
    await client.acreate_bank(bank_id="transfer-py-other")

    # [docs:prompts-preview]
    from hindsight_client_api.models import PromptPreviewRequest

    preview = await client.banks.preview_prompt("my-bank", PromptPreviewRequest(operation="retain"))
    for message in preview.messages:
        print(message.role, len(message.blocks))
    # [/docs:prompts-preview]
    assert preview.messages

    # [docs:transfer-export]
    # Whole bank, memories + config, no history.
    # Submits the export, polls the operation, downloads the ZIP.
    archive = await client.aexport_bank("transfer-py")

    # Just the memories
    memories_only = await client.aexport_bank("transfer-py", include_bank_config=False)

    # Specific documents (a document subset carries no bank-level sections).
    # The low-level call only submits; poll the returned operation yourself.
    submission = await client.bank_transfer.export_bank_transfer(
        "transfer-py", document_id=["doc-1", "doc-2"], include_bank_config=False
    )
    # [/docs:transfer-export]
    assert archive[:2] == b"PK" and memories_only[:2] == b"PK"
    await wait_for(client, "transfer-py", submission.operation_id)

    # [docs:transfer-import]
    # Restore a bank under a new id
    operation_id = await client.aimport_bank("transfer-py", archive, target_bank_id="transfer-py-copy")
    # The restore is recorded against the bank in the URL — poll it there
    status = await client.operations.get_operation_status("transfer-py", operation_id)

    # Merge an archive's documents into an existing bank
    submission = await client.bank_transfer.import_bank_transfer(
        "transfer-py-other",
        ("transfer-py.zip", archive),
        mode="merge",
        document_conflict="replace",
    )
    # [/docs:transfer-import]
    await wait_for(client, "transfer-py", operation_id)
    await wait_for(client, "transfer-py-other", submission.operation_id)
    copied = await client.documents.list_documents("transfer-py-copy")
    assert copied.total == 2, copied

    # [docs:clone-bank]
    operation_id = await client.aclone_bank("transfer-py", "transfer-py-clone")
    # The operation is recorded against the source bank
    status = await client.operations.get_operation_status("transfer-py", operation_id)
    # [/docs:clone-bank]
    await wait_for(client, "transfer-py", operation_id)
    cloned = await client.documents.list_documents("transfer-py-clone")
    assert cloned.total == 2, cloned

    # [docs:document-export]
    # Submits the export (whole bank; pass document_ids=[...] to scope it),
    # polls the operation until completed, downloads the archive.
    archive = await client.aexport_documents("transfer-py")
    with open("transfer-py-documents.zip", "wb") as f:
        f.write(archive)
    # [/docs:document-export]
    assert archive[:2] == b"PK"

    # [docs:document-import]
    with open("transfer-py-documents.zip", "rb") as f:
        submission = await client.document_transfer.import_documents(
            "transfer-py-other", ("transfer-py-documents.zip", f.read()), on_conflict="replace"
        )

    status = await client.operations.get_operation_status("transfer-py-other", submission.operation_id)
    # status.result_metadata -> {"documents_imported": 3, "facts_imported": 42, "observations_imported": 5, ...}
    # [/docs:document-import]
    await wait_for(client, "transfer-py-other", submission.operation_id)
    os.remove("transfer-py-documents.zip")

    for bank_id in TRANSFER_BANKS:
        requests.delete(f"{HINDSIGHT_URL}/v1/default/banks/{bank_id}")
    await client.aclose()


asyncio.run(transfer_examples())


# =============================================================================
# Cleanup (not shown in docs)
# =============================================================================
requests.delete(f"{HINDSIGHT_URL}/v1/default/banks/my-bank")
requests.delete(f"{HINDSIGHT_URL}/v1/default/banks/architect-bank")

print("memory-banks.py: All examples passed")

File: hindsight-docs/examples/api/memory-banks.mjs
#!/usr/bin/env node
/**
 * Memory Banks API examples for Hindsight (Node.js)
 * Run: node examples/api/memory-banks.mjs
 */
import { HindsightClient, sdk, createClient, createConfig } from '@vectorize-io/hindsight-client';
import { readFile, writeFile, rm } from 'node:fs/promises';

const HINDSIGHT_URL = process.env.HINDSIGHT_API_URL || 'http://localhost:8888';

// =============================================================================
// Setup (not shown in docs)
// =============================================================================
const client = new HindsightClient({ baseUrl: HINDSIGHT_URL });

// =============================================================================
// Doc Examples
// =============================================================================

// [docs:create-bank]
await client.createBank('my-bank');
// [/docs:create-bank]


// [docs:bank-with-disposition]
await client.createBank('architect-bank');
await client.updateBankConfig('architect-bank', {
    reflectMission: "You're a senior software architect - keep track of system designs, technology decisions, and architectural patterns.",
    dispositionSkepticism: 4,   // Questions new technologies
    dispositionLiteralism: 4,   // Focuses on concrete specs
    dispositionEmpathy: 2,      // Prioritizes technical facts
});
// [/docs:bank-with-disposition]


// [docs:bank-background]
await client.createBank('my-bank');
await client.updateBankConfig('my-bank', {
    reflectMission: 'I am a research assistant specializing in machine learning.',
});
// [/docs:bank-background]


// [docs:bank-mission]
await client.createBank('my-bank');
await client.updateBankConfig('my-bank', {
    reflectMission: "You're a senior software architect - keep track of system designs, technology decisions, and architectural patterns.",
});
// [/docs:bank-mission]


// [docs:update-bank-config]
await client.updateBankConfig('my-bank', {
    retainMission: 'Always include technical decisions, API design choices, and architectural trade-offs. Ignore meeting logistics and social exchanges.',
    retainExtractionMode: 'verbose',
    observationsMission: 'Observations are stable facts about people and projects. Always include preferences, skills, and recurring patterns. Ignore one-off events.',
    dispositionSkepticism: 4,
    dispositionLiteralism: 4,
    dispositionEmpathy: 2,
});
// [/docs:update-bank-config]


// [docs:get-bank-config]
// Returns resolved config (server defaults merged with bank overrides) and the raw overrides
const { config, overrides } = await client.getBankConfig('my-bank');
// config    — full resolved configuration
// overrides — only fields overridden at the bank level
// [/docs:get-bank-config]


// [docs:reset-bank-config]
// Remove all bank-level overrides, reverting to server defaults
await client.resetBankConfig('my-bank');
// [/docs:reset-bank-config]


// =============================================================================
// Prompt preview and bank transfer
// =============================================================================
const apiClient = createClient(createConfig({ baseUrl: HINDSIGHT_URL }));
const TRANSFER_BANKS = ['transfer-js', 'transfer-js-copy', 'transfer-js-other', 'transfer-js-clone'];
for (const b of TRANSFER_BANKS) await fetch(`${HINDSIGHT_URL}/v1/default/banks/${b}`, { method: 'DELETE' });
// Chunks mode stores text verbatim, so no LLM call is needed.
await client.createBank('transfer-js', {});
await client.updateBankConfig('transfer-js', { retainExtractionMode: 'chunks' });
const retainRes = await fetch(`${HINDSIGHT_URL}/v1/default/banks/transfer-js/memories`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ items: [
        { content: 'Alice leads the payments team.', document_id: 'doc-1' },
        { content: 'Bob moved to the Berlin office.', document_id: 'doc-2' },
    ] }),
});
if (!retainRes.ok) throw new Error(`retain failed: ${retainRes.status}`);
await client.createBank('transfer-js-other', {});

async function waitFor(bankId, operationId) {
    for (let i = 0; i < 120; i++) {
        const { data } = await sdk.getOperationStatus({
            client: apiClient, path: { bank_id: bankId, operation_id: operationId },
        });
        if (data.status === 'completed') return data;
        if (data.status === 'failed' || data.status === 'cancelled') {
            throw new Error(`operation ${operationId} ${data.status}: ${data.error_message}`);
        }
        await new Promise((r) => setTimeout(r, 1000));
    }
    throw new Error(`operation ${operationId} did not finish`);
}

// [docs:prompts-preview]
const { data: preview } = await sdk.previewPrompt({
    client: apiClient,
    path: { bank_id: 'my-bank' },
    body: { operation: 'retain' },
});
for (const message of preview.messages) console.log(message.role, message.blocks.length);
// [/docs:prompts-preview]
if (!preview.messages.length) throw new Error('empty preview');


// [docs:transfer-export]
// Whole bank, memories + config, no history.
// Submits the export, polls the operation, downloads the ZIP.
const archive = await client.exportBank('transfer-js');

// Just the memories
const memoriesOnly = await client.exportBank('transfer-js', { includeBankConfig: false });

// Specific documents (a document subset carries no bank-level sections).
// The low-level call only submits; poll the returned operation yourself.
const { data: subset } = await sdk.exportBankTransfer({
    client: apiClient,
    path: { bank_id: 'transfer-js' },
    query: { document_id: ['doc-1', 'doc-2'], include_bank_config: false },
});
// [/docs:transfer-export]
if (archive[0] !== 0x50 || memoriesOnly[0] !== 0x50) throw new Error('not a zip');
await waitFor('transfer-js', subset.operation_id);


// [docs:transfer-import]
// Restore a bank under a new id
const restoreId = await client.importBank('transfer-js', new Blob([archive]), {
    targetBankId: 'transfer-js-copy',
});
// The restore is recorded against the bank in the URL — poll it there
const { data: restoreStatus } = await sdk.getOperationStatus({
    client: apiClient,
    path: { bank_id: 'transfer-js', operation_id: restoreId },
});

// Merge an archive's documents into an existing bank
const { data: merge } = await sdk.importBankTransfer({
    client: apiClient,
    path: { bank_id: 'transfer-js-other' },
    query: { mode: 'merge', document_conflict: 'replace' },
    body: { file: new Blob([archive]) },
});
// [/docs:transfer-import]
await waitFor('transfer-js', restoreId);
await waitFor('transfer-js-other', merge.operation_id);
const copied = await client.listDocuments('transfer-js-copy');
if (copied.total !== 2) throw new Error(`restore copied ${copied.total} documents`);


// [docs:clone-bank]
const cloneId = await client.cloneBank('transfer-js', 'transfer-js-clone');
// The operation is recorded against the source bank
const { data: cloneStatus } = await sdk.getOperationStatus({
    client: apiClient,
    path: { bank_id: 'transfer-js', operation_id: cloneId },
});
// [/docs:clone-bank]
await waitFor('transfer-js', cloneId);
const cloned = await client.listDocuments('transfer-js-clone');
if (cloned.total !== 2) throw new Error(`clone copied ${cloned.total} documents`);


// [docs:document-export]
// Submits the export (whole bank; pass { documentIds: [...] } to scope it),
// polls the operation until completed, downloads the archive.
const docArchive = await client.exportDocuments('transfer-js');
await writeFile('transfer-js-documents.zip', docArchive);
// [/docs:document-export]


// [docs:document-import]
const { data: docImport } = await sdk.importDocuments({
    client: apiClient,
    path: { bank_id: 'transfer-js-other' },
    query: { on_conflict: 'replace' },
    body: { file: new Blob([await readFile('transfer-js-documents.zip')]) },
});

const { data: docImportStatus } = await sdk.getOperationStatus({
    client: apiClient,
    path: { bank_id: 'transfer-js-other', operation_id: docImport.operation_id },
});
// docImportStatus.result_metadata -> { documents_imported: 3, facts_imported: 42, observations_imported: 5, ... }
// [/docs:document-import]
await waitFor('transfer-js-other', docImport.operation_id);
await rm('transfer-js-documents.zip');

// [docs:transfer-import-external]
import { crc32 } from 'node:zlib';

// Minimal uncompressed ZIP writer (Node has none built in); a library like jszip works too.
function zip(files) {
    const locals = [], centrals = [];
    let offset = 0;
    for (const [name, text] of Object.entries(files)) {
        const n = Buffer.from(name), d = Buffer.from(text), crc = crc32(d);
        const local = Buffer.alloc(30);
        local.writeUInt32LE(0x04034b50, 0); local.writeUInt16LE(20, 4); local.writeUInt16LE(0x21, 12);
        local.writeUInt32LE(crc, 14); local.writeUInt32LE(d.length, 18); local.writeUInt32LE(d.length, 22);
        local.writeUInt16LE(n.length, 26);
        const central = Buffer.alloc(46);
        central.writeUInt32LE(0x02014b50, 0); central.writeUInt16LE(20, 4); central.writeUInt16LE(20, 6);
        central.writeUInt16LE(0x21, 14); central.writeUInt32LE(crc, 16); central.writeUInt32LE(d.length, 20);
        central.writeUInt32LE(d.length, 24); central.writeUInt16LE(n.length, 28); central.writeUInt32LE(offset, 42);
        locals.push(local, n, d);
        centrals.push(central, n);
        offset += 30 + n.length + d.length;
    }
    const dir = Buffer.concat(centrals), end = Buffer.alloc(22);
    const count = Object.keys(files).length;
    end.writeUInt32LE(0x06054b50, 0); end.writeUInt16LE(count, 8); end.writeUInt16LE(count, 10);
    end.writeUInt32LE(dir.length, 12); end.writeUInt32LE(offset, 16);
    return Buffer.concat([...locals, dir, end]);
}

const doc = {
    id: 'session-2026-09-22',
    original_text: 'Full original session text...',
    chunks: [{ chunk_index: 0, chunk_text: 'Caller-defined source region...' }],
    facts: [{
        text: 'The user prefers lightweight local speech recognition models.',
        fact_type: 'experience',
        chunk_index: 0,
        mentioned_at: '2026-09-22T18:34:00Z',
        entities: ['Parakeet'],
    }],
};
const archiveZip = zip({
    'manifest.json': JSON.stringify({ schema_version: 1, source_bank_id: 'external' }),
    [`documents/${doc.id}.json`]: JSON.stringify(doc),
});

const { data: external } = await sdk.importBankTransfer({
    client: apiClient,
    path: { bank_id: 'transfer-js-other' },
    query: { mode: 'merge', document_conflict: 'replace' },
    body: { file: new Blob([archiveZip]) },
});
// [/docs:transfer-import-external]
await waitFor('transfer-js-other', external.operation_id);
const { data: importedDoc } = await sdk.getDocument({
    client: apiClient,
    path: { bank_id: 'transfer-js-other', document_id: 'session-2026-09-22' },
});
if (importedDoc.memory_unit_count !== 1) throw new Error(`external import stored ${importedDoc.memory_unit_count} facts`);
for (const b of TRANSFER_BANKS) await fetch(`${HINDSIGHT_URL}/v1/default/banks/${b}`, { method: 'DELETE' });


// =============================================================================
// Cleanup (not shown in docs)
// =============================================================================
await fetch(`${HINDSIGHT_URL}/v1/default/banks/my-bank`, { method: 'DELETE' });
await fetch(`${HINDSIGHT_URL}/v1/default/banks/architect-bank`, { method: 'DELETE' });

console.log('memory-banks.mjs: All examples passed');
